vpr/src/device/rr_spatial_lookup.h

```diff
@@ -0,0 +1,97 @@
+#ifndef RR_SPATIAL_LOOKUP_H
+#define RR_SPATIAL_LOOKUP_H
+
+#include "vpr_types.h"
+
+/******************************************************************** 
+ * A data structure storing the fast look-up for the nodes
+ * in a routing resource graph
+ *
+ * The data structure allows users to 
+ * - Update the look-up with new nodes
+ * - Find the id of a node with given information, e.g., x, y, type etc.
+ ********************************************************************/
+class RRSpatialLookup {
+    /* -- Constructors -- */
+  public:
+    /* Explicitly define the only way to create an object */
+    explicit RRSpatialLookup(t_rr_node_indices& rr_node_indices);
+
+    /* Disable copy constructors and copy assignment operator
+     * This is to avoid any duplication of the object
+     * as it is only interface allowed to access node look-up of a routing resource graph
+     */
+    RRSpatialLookup(const RRSpatialLookup&) = delete;
+    void operator=(const RRSpatialLookup&) = delete;
+
+    /* -- Accessors -- */
+  public:
+    /* Returns the index of the specified routing resource node.  (x,y) are
+     * the location within the FPGA, rr_type specifies the type of resource,
+     * and ptc gives the number of this resource.  ptc is the class number,
+     * pin number or track number, depending on what type of resource this
+     * is.  All ptcs start at 0 and go up to pins_per_clb-1 or the equivalent.
+     * There are class_inf size SOURCEs + SINKs, type->num_pins IPINs + OPINs,
+     * and max_chan_width CHANX and CHANY (each).
+     *
+     * Note that for segments (CHANX and CHANY) of length > 1, the segment is
+     * given an rr_index based on the (x,y) location at which it starts (i.e.
+     * lowest (x,y) location at which this segment exists).
+     * This routine also performs error checking to make sure the node in
+     * question exists.
+     *
+     * The 'side' argument only applies to IPIN/OPIN types, and specifies which
+     * side of the grid tile the node should be located on. The value is ignored
+     * for non-IPIN/OPIN types
+     */
+    RRNodeId find_node(int x,
+                       int y,
+                       t_rr_type type,
+                       int ptc,
+                       e_side side = NUM_SIDES) const;
+
+    /* -- Mutators -- */
+  public:
+    /* Register a node in the fast look-up 
+     * - You must have a node id
+     *   1. a valid one is to register a node in the lookup
+     *   2. an invalid id means the removal of a node from the lookup
```

Comment:
You are right. The use of invalid id is very tricky. It just unregisters a node from the lookup.
Actually, I should not state it because we should not promote such use of the API.
Therefore, I removed this line. We should add an API ``remove_node()``. I put this TODO item in the comment.